src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -823,47 +810,63 @@ private ImmutableArray<Symbol> CreateMembers()
+                                // if provided name does not match the default one, 
+                                // then default element as declared implicitly
+                                var defaultImplicitlyDeclared = providedName != defaultName;
+
                                 var fieldSymbol = field.AsMember(currentUnderlying);
 
                                 if (currentNestingLevel != 0)
                                 {
                                     // This is a matching field, but it is in the extension tuple
-                                    tupleFieldIndex += (RestPosition - 1) * currentNestingLevel;
 
-                                    string defaultName = TupleMemberName(tupleFieldIndex + 1);
-                                    // Add a field with default name if the given name is different
-                                    if (namesOfVirtualFields[tupleFieldIndex] != defaultName)
+                                    // Add a field with default name. It should be present regardless.
+                                    // Make it virtual since we are not at the top level
+                                    // tupleFieldIndex << 1 because this is a default element
+                                    members.Add(new TupleVirtualElementFieldSymbol(this, fieldSymbol, defaultName, tupleFieldIndex << 1, location, defaultImplicitlyDeclared));
+
+                                    if (defaultImplicitlyDeclared && providedName != null)
                                     {
                                         // The name given doesn't match default name Item8, etc.
-                                        // Add a field with default name and make it virtual since we are not at the top level
-                                        members.Add(new TupleVirtualElementFieldSymbol(this, fieldSymbol, defaultName, tupleFieldIndex, null));
+                                        // Add a field with the given name
+                                        // tupleFieldIndex << 1 + 1, because this is not a default element
+                                        members.Add(new TupleVirtualElementFieldSymbol(this, fieldSymbol, providedName, (tupleFieldIndex << 1) + 1, location, isImplicitlyDeclared: false));
                                     }
-
-                                    // Add a field with the given name
-                                    var location = _elementLocations.IsDefault ? null : _elementLocations[tupleFieldIndex];
-                                    members.Add(new TupleVirtualElementFieldSymbol(this, fieldSymbol, namesOfVirtualFields[tupleFieldIndex],
-                                                        tupleFieldIndex, location));
-
                                 }
-                                else if (field.Name == namesOfVirtualFields[tupleFieldIndex])
+                                else if (defaultImplicitlyDeclared)
```

Comment:
Merging the branches will, indeed, make the code more compact, but I am afraid it will be harder to read. 
I think, the logic is too subtle and having all cases handled explicitly allows for better placing of comments and asserts. It also allows the reader to concentrate on each of the cases separately.

Basically, I think more lines of code here is a good thing and we can tolerate some redundancy for readability purposes.